### ISSUE TITLE
Difference problem between Local explainability and Global explainability

### ISSUE THREAD

**Issue by yaoching0** [dreamquark-ai/tabnet#371]
HI, 
I train tabnet with 4700-dimension feature,and i check the Global explainability.(Because of its sparseness, I did unique(),the s shows the index in the Global explainability matrix )
and i input training data to the function .explain(Training data), i sum explain_matrix cross the rows, but i get totally different result with Global explainability.
For example,the index of the globally explained maximum value is even 0 in the output of explain() function
![1647936738488](https://user-images.githubusercontent.com/96738701/159436232-9eb84448-5506-4acc-a4c3-d4348f02c180.jpg)

**Comment by eduardocarvp** [comment]
Hello @yaoching0 ,

Are you using embeddings for categorical features on your model?

**Comment by yaoching0** [comment]
@eduardocarvp No, all numerical data between (0,1)

**Comment by yaoching0** [comment]
My data format is processed and trained exactly by example.
![1647941394526](https://user-images.githubusercontent.com/96738701/159449639-1ea6d0e5-2040-4c86-9ee3-e8a512d17b2b.jpg)
![1647941412820](https://user-images.githubusercontent.com/96738701/159449655-dbec3cc2-c354-4f22-8267-f69afbdac02b.jpg)

**Comment by yaoching0** [comment]
This is my training data format
![1647942053845](https://user-images.githubusercontent.com/96738701/159451677-81c9ab4d-a2bf-48c0-9b35-6a54867230a0.jpg)

**Comment by Optimox** [comment]
The `clf.feature_importances` are normalized while the individual importance are not, have a look at #180 

So you need to divide by the sum of each row. Moreover, you want to average no sum
`avg_imp = np.mean(explain_matrix, axis=0)
avg_imp = avg_imp / np.sum(avg_imp)
`
(make sure that axes are correct)

**Comment by yaoching0** [comment]
@Optimox thanks for reply.
I already know the individual importance are not normalized.
but the index of max value in clf.feature_importances  becomes **0** in the same index of the explain_matrix i sum across the rows.
I think it shouldn't be zero even though I'm not dividing by the sum of each row of explain_matrix
![1647943189810](https://user-images.githubusercontent.com/96738701/159455371-a5bc82a9-b79b-4507-abc4-678c3b69412e.jpg)

**Comment by yaoching0** [comment]
They still don't match
![1647943667910](https://user-images.githubusercontent.com/96738701/159456820-1a02a41d-9bcc-42b0-a8f4-8a6c3d6885d1.jpg)

**Comment by yaoching0** [comment]
@Optimox 
I have a new clue and I see that after calculation they have the same value, but are indeed in different positions, why is this happening?
![1647947051667](https://user-images.githubusercontent.com/96738701/159467876-faf525b9-7c11-4d70-a6d6-0c9596f32edd.jpg)

**Comment by yaoching0** [comment]
I restart the jupyter kernel,and is seems match

**Comment by Optimox** [comment]
@yaoching0,

Actually I'm surprised that things are working now for you.

I did a test and I was not able to find exactly the `feature_importance`, this is due to the fact that the internal `feature_importance` are computed on the train DATALOADER while it creates a new dataloader when calling `clf.explain(X)`.

So if you have any parameters that changes the train dataloader you'll end up with a different score. This can happen in different scenarios:
- `drop_last=True` : you are computing the feature importance without a few lines (randomly selected as shuffle=True for training dataloader). Potentially this means that the feature importance is somehow random (but still representative). I think it's ok to call this a bug.
- 'weights=1` : you are going to over-sample some examples in your training loader, which will change the final feature_importance.

I think those are the only two reasons, but there might be a few other scenarios that I did not spotted.

In the end: this is a bug, I'll fix it, thank you very much for finding it. In the meantime, the differences are coming from the sample used for internal feature importance. So you can trust both methods: if you see a big change in your case, this is due to the high sparsity of your data, and the final training importance might not be very representative.

**Comment by eduardocarvp** [comment]
Yes, I think there are some problems as well. I was wondering if the reducing matrix has any influence, since in one case we do the sum first and then reduction and the inverse in the other case. But since there are no np.abs and only sums and averages I guess it should be the same...
